### PR TITLE
Exec on dst and esdt call

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

**Review by andrei-marinica — APPROVED** [review]

**Review by andrei-marinica — DISMISSED** [review]

File: arwen/host/execution.go [hunk 1]
```diff
@@ -655,12 +659,15 @@ func (host *vmHost) RevertESDTTransfer(input *vmcommon.ContractCallInput) {
...
+	copy(revertInput.Arguments, input.Arguments)
 	if input.Function == core.BuiltInFunctionESDTNFTTransfer {
-		revertInput.RecipientAddr = revertInput.CallerAddr
-		// in esdt nft transfer the 4th arguments is actually the destination address
-		revertInput.Arguments[numArgsForTransfer-1] = input.CallerAddr
+		actualRecipient := input.CallerAddr
+		actualSender := input.Arguments[numArgsForTransfer-1]
+
+		revertInput.RecipientAddr = actualSender
+		revertInput.CallerAddr = actualSender
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go:668]: NFT transfers read the recipient only from the arguments?

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -2271,6 +2435,7 @@ func prepareIndirectContractCallInput(
 	numArguments int32,
 	argumentsLengthOffset int32,
 	dataOffset int32,
+	syncExecution bool,
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:2438]: Suggest renaming this parameter to `syncExecutionRequired`, it appears to be a better name given its usage in the `if` below.

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -1939,6 +2062,45 @@ func executeOnDestContext(
 		return 1
 	}
 
+	if host.IsESDTFunctionsEnabled() {
+		function, err := runtime.MemLoad(functionOffset, functionLength)
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:2066]: Please create the function `host.detectSyncESDTTransfer(functionOffset, functionLength, addressOffset)` returning `function, destination, isESDTTransfer, err` to simplify this code.

File: arwen/host/execution.go [hunk 4]
```diff
@@ -206,12 +206,16 @@ func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (vmO
 	defer func() {
 		vmOutput = host.finishExecuteOnDestContext(err)
 		metering.SetTotalUsedGas(0)
+
+		if err == nil && vmOutput.ReturnCode != vmcommon.Ok {
+			err = arwen.ErrExecutionFailed
+		}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/host/execution.go:212]: it is necessary if we make an error on gas usage in arwen. Thus finishExecuteOnDestContext can create vmOutput with error if gas usage is missmatched

File: arwen/contexts/output.go [hunk 5]
```diff
@@ -271,8 +271,9 @@ func (context *outputContext) TransferValueOnly(destination []byte, sender []byt
 	}
 
 	isAsyncCall := context.host.IsArwenV3Enabled() && context.host.Runtime().GetVMInput().CallType == vmcommon.AsynchronousCall
-	hasValue := value.Cmp(big.NewInt(0)) == 1
-	if !payable && hasValue && !isAsyncCall {
+	skipPayableCheck := context.host.IsESDTFunctionsEnabled() && !checkPayable
+	hasValue := value.Cmp(arwen.Zero) > 0
+	if !payable && hasValue && !isAsyncCall && !skipPayableCheck {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/output.go:276]: done

File: arwen/contexts/output.go [hunk 6]
```diff
@@ -290,7 +291,7 @@ func (context *outputContext) TransferValueOnly(destination []byte, sender []byt
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (context *outputContext) Transfer(destination []byte, sender []byte, gasLimit uint64, gasLocked uint64, value *big.Int, input []byte, callType vmcommon.CallType) error {
-	err := context.TransferValueOnly(destination, sender, value)
+	err := context.TransferValueOnly(destination, sender, value, gasLimit > 0 && callType != vmcommon.AsynchronousCallBack)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/output.go:294]: done

File: arwen/host/execution.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/host/execution.go:668]: this is ok.

File: arwen/elrondapi/elrondei.go (hunk 2 above)
Comment (sasurobert) [reply arwen/elrondapi/elrondei.go:2438]: done

File: arwen/elrondapi/elrondei.go (hunk 3 above)
Comment (sasurobert) [reply arwen/elrondapi/elrondei.go:2066]: done

**Review by andrei-marinica — APPROVED** [review]

**Review by camilbancioiu — APPROVED** [review]

**Pull request merged**